calculatorFrame.java

```diff
@@ -111,191 +104,192 @@ public void init() {
 		clearButtonPanel.add(clearAll);
 		content.add(clearButtonPanel,BorderLayout.NORTH);
 	}
-	
+
 	public void actionPerformed(ActionEvent e) {
-		
+
 		//One button
-		if(e.getSource() == one){
-			if (flag == false){
+		if(e.getSource() == one) {
```

Comment:
It looks like there’s a lot of duplicated code between the buttons `zero` through `nine`, which is a code smell – it makes me think that there’s some shared behaviour between the buttons that needs to live in its own method.

To figure out how/where that method should live, try asking yourself “what do I wish I could I tell my objects to do?” In this case, you’re checking `e.getSource()` to see if it’s a number button, and then you have logic (printing and setting text) and some data (the "1" string).

What if you could have something like:

```
public class JNumberButton extends JButton {
  public double number;

  public void init(newNumber) {
    number = newNumber;
    super(number.toString());
  }
}

one = new JNumberButton(1);
two = new JNumberButton(2);
...

JNumberButton numberButton;
if(numberButton = isNumberButton(e.getSource()) {
  System.out.println(firstNumString);
  firstNumString += numberButton.number.toString();
  System.out.println(firstNumString);
  textArea.setText(textArea.getText() + e.getActionCommand());
}
```